readme.md
# Commander x16  development tools

![Writing code title](images/writingcodecropped200.png)

## Commander x16 Website

[Visit Commander x16 Website](https://www.commanderx16.com/)

<a href="https://www.buymeacoffee.com/cavtronics" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" style="height: 60px !important;width: 217px !important;" ></a>

 <a href="https://www.buymeacoffee.com/cavtronics" title="Donate to my libraries using BuyMeACoffee"><img src="https://img.shields.io/badge/buy%20me%20a%20coffee-donate-orange.svg?logo=buy-me-a-coffee&logoColor=FFDD00" style="height: 20px !important;width: 200px !important;" ></a>

[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](https://makeapullrequest.com)

## Version

current version v0.6.0

## Introduction

This is a development environment for Commander x16 and Commodore 64 computers.
It is assumed that you have Visual Studio Code installed.

## Changelog

check the changelog.txt for history and list of changes

## Using the tools and Installation

Read the Instructions.md for how-to configure and use.

## Contributing

If you found an error, a typo or have an idea and wish to contribute
read the Contributing.md file

## Cloning

To clone create a subdirectory then change to that directory and use the following command from the command prompt:

        git clone  https://github.com/pacav69/x16devproject

this will copy the files to your computer into the created subdirectory.

## Tools used

* Visual Studio Code
<img src="https://visualstudio.microsoft.com/wp-content/uploads/2019/09/vs-code-responsive-01-1.png" alt="Visual Studio Code logo" style="height: 50px; width:60px;"/>

Visual Studio Code Editor - [found here](https://code.visualstudio.com/)

* ACME Cross assembler [found here](https://csdb.dk/release/?id=193302)
* ACME Cross assembler on github [here](https://github.com/meonwax/acme)

## Videos and tutorials

* Using the Commander X16 Integrated Assembly Environment - Codex [here](https://www.youtube.com/watch?v=N3E0MUsqzQo&ab_channel=JestinStoffel)
* ChibiAkumas  [here](https://www.youtube.com/@ChibiAkumas)
* ChibiAkumas website [here](https://www.assemblytutorial.com/)
* TheRetroDesk [here](https://www.youtube.com/watch?v=gqwIzbT_WFU&ab_channel=TheRetroDesk)
* TheRetroDesk github files  [here](https://github.com/SlithyMatt/x16-assembly-tutorial)
* Commander X16 Assembly Project Setup [here](https://www.youtube.com/watch?v=V2vGiIqDXkA&ab_channel=JestinStoffel)
* Commander X16 6502 Assembly Language Dev Setup [here](https://www.youtube.com/watch?v=oz3E6pGU2KE&ab_channel=Spriteworx)
* Commander X16 6502 Assembly Language Dev Setup by Spriteworx [here](https://www.youtube.com/watch?v=oz3E6pGU2KE&ab_channel=Spriteworx)
* Assembly in Basic - Setting up the environment [here](https://www.8bitcoding.com/p/assembly-in-basic-i-setting-up.html)

## Visual Studio Code extensions

* VS64 - The C64 Development Environment  [here](https://marketplace.visualstudio.com/items?itemName=rosc.vs64)
* C64 [here](https://marketplace.visualstudio.com/items?itemName=gverduci.c64basicv2)
* kick-assembler [here](https://marketplace.visualstudio.com/items?itemName=paulhocker.kick-assembler-vscode-ext)
* Some VSCode extensions in 2023 video [here](https://www.youtube.com/watch?v=DNf6Bu7z4vw&ab_channel=DevtipsbyMoHo)

### Commander x16 links

* reference  manual [here](https://github.com/x16community/x16-docs)
* forum [here](https://www.commanderx16.com/forum)
* Commander x16 emulator [here](https://github.com/x16community/x16-emulator)
* retro-code  gitlab [here](https://gitlab.com/retro-coder/commanderx16)

### Data sheets

* [W65C02](https://eater.net/datasheets/w65c02s.pdf)
* CPU w65c02 datasheet [here](https://westerndesigncenter.com/wdc/documentation/w65c02s.pdf)

* [W65C22 (VIA)](https://eater.net/datasheets/w65c22.pdf)
* [W65C51 (ACIA)](https://eater.net/datasheets/w65c51n.pdf)

## C64 References

* Existing reference material [here](https://github.com/X16Community/x16-docs/issues/83)
* cbm docs [here](https://sta.c64.org/cbmdocs.html)
* C64 collections [here](https://www.lyonlabs.org/commodore/onrequest/collections.html)
* hello world example [here](https://retro-programming.com/hello-world-in-the-c64-machine-code/)

* cbm64 kernal function  [here](http://sta.c64.org/cbm64krnfunc.html)
* c64 kernal ROM  [here](https://c64os.com/post/c64kernalrom) pretty good writeup with explanations
* c64 reference  [here](https://www.pagetable.com/c64ref/kernal/)
* C64 Kernal routines [here](https://cx16.dk/c64-kernal-routines/)
* C64 colors [here](https://www.c64-wiki.com/wiki/Color)
* Petmate is a cross-platform C64 PETSCII image editor. [here](https://nurpax.github.io/petmate/)

Here's some more, courtesy of @JimmyDansbo :

* c64 kernal routines [here](https://www1.cx16.dk/c64-kernal-routines/)
* c128-kernal-routines [here](https://www1.cx16.dk/c128-kernal-routines/)
* 6502opcodes [here](http://6502.org/tutorials/6502opcodes.html)
* 65c02opcodes [here](http://www.6502.org/tutorials/65c02opcodes.html)

Maybe less useful on the x16:

* C64 dissembler [here](https://www.pagetable.com/c64ref/c64disasm/)
* map c6411 [here](https://project64.c64.org/Software/mapc6411.txt)

### Assemblers

* Acme cross assembler for windows [here](https://sourceforge.net/projects/acme-crossass/files/win32/acme0.97win.zip/download)
* cc65 compiler [here](https://cc65.github.io/)
* KickAssembler main page [here]( http://theweb.dk/KickAssembler/Main.html#frontpage)
* KickAssembler zip file [here](http://www.theweb.dk/KickAssembler/KickAssembler.zip)
* retro assembler [here](https://enginedesigns.net/retroassembler/)
* Download .NET 6.0 support for retro assembler [here](https://dotnet.microsoft.com/en-us/download/dotnet/6.0)
* acme on github [here](https://github.com/meonwax/acme)

### Emulators

* Commander x16 [here](https://github.com/x16community/x16-emulator)
* SNES emulator (FCEUX) [here](https://fceux.com/web/home.html)
* BizHawk is a multi-platform emulator  [here](https://tasvideos.org/Bizhawk)

### Visual studio code

vscode variables-reference [here](https://code.visualstudio.com/docs/editor/variables-reference)

## Support

<img src="https://vangogh.teespring.com/v3/image/SugZ-DRGZXUTuSzfrFtaOU3TAUQ/800/800.jpg" width="100px"  height="100px">

[Help support by buying some merchandise](https://cavtronics-3.creator-spring.com/)

## Acknowledgements


changelog.txt
changelog.txt

---------------------------------------------
04Dec2023
---------------------------------------------
added c64opcodes extension
added box comment extension

---------------------------------------------
26Nov2023
---------------------------------------------
added video link for codex

---------------------------------------------
22Nov2023
---------------------------------------------
added to task acme compile and run emulator
added reference to Petmate is a cross-platform C64 PETSCII image editor.
added reference to vscode variables-reference handy for tasks.json

---------------------------------------------
17Nov2023
---------------------------------------------
added new file: Debugger.md
added video link Commander X16 - Using the 6502 Assembly Dev Tools

---------------------------------------------
16Nov2023
---------------------------------------------
added using instructions for retro assembler
added helloworldretro.65c02.asm file for testing of retro assembler

---------------------------------------------
15Nov2023
---------------------------------------------
added retro assembler requires https://dotnet.microsoft.com/en-us/download/dotnet/6.0
updated x16 emulator to r46

---------------------------------------------
08Nov2023
---------------------------------------------
updated readme
updated instructions
added readme.txt
added hello world example

---------------------------------------------
06Nov2023
---------------------------------------------
updated readme file
added Contributing.md
added writing code image
updated instructions.md
added workspaacerecommended.png


---------------------------------------------
05Nov2023
---------------------------------------------
Initial
updated instructions
added cx16-maze.asm as an example by Jimmy Dansbo found here https://github.com/JimmyDansbo/cx16-maze

src/5heartsg.asm
;   ┌──────────────────────────────────────────────────────────────────────────┐
;   │ 5 Green hearts                                                           │
;   ├──────────────────────────────────────────────────────────────────────────┤
;   │ by                                                                       │
;   └──────────────────────────────────────────────────────────────────────────┘
*=$0801
!byte $01,$08,$0b,$08,$01,$00,$9e,$32,$30,$36,$31,$00,$00,$00

@ #0
STA $9F25	; Select primary VRAM address
LDA #$20	; VPOKE 1st argument (The 0x00 in this is the 0 bank)
STA $9F22	; Set primary address bank to 0, stride to 2
LDA #0
STA $9F21	; Set primary address high byte to 0

LDX #0		; Loop counter
LDY #0		; Address offset
next_heart:
	TYA
	STA $9F20	; Set Primary address low byte to 0

	LDA #$53	; <3
	STA $9F23	; Data line
	INY
	TYA		; Next byte over is color
	STA $9f20
	LDA #5		; Green
	STA $9f23	; Write the color
	INX
	INY
	CPX #5
	BNE next_heart
BRK
	ORA
	AND
	BBR
	BNE
	RMB0

	ADC
	AAA


src/helloworld.asm
; Hello world example
; https://github.com/JimmyDansbo/cx16stuff/blob/master/helloworld.asm
; designed for compiling with ACME assembler
!cpu 65c02		; set target processor
*=$0801				; Assembled code should start at $0801
				; (where BASIC programs start)
				; The real program starts at $0810 = 2064
!byte $0C,$08			; $080C - pointer to next line of BASIC code
!byte $0A,$00			; 2-byte line number ($000A = 10)
!byte $9E			; SYS BASIC token
!byte $20			; [space]
!byte $32,$30,$36,$34		; $32="2",$30="0",$36="6",$34="4"
				; (ASCII encoded nums for dec starting addr)
!byte $00			; End of Line
!byte $00,$00			; This is address $080C containing
				; 2-byte pointer to next line of BASIC code
				; ($0000 = end of program)
*=$0810				; Here starts the real program

CHROUT=$FFD2			; CHROUT outputs a character (C64 Kernal API)
CHRIN=$FFCF			; CHRIN read from default input

	ldx	#0		; X register is used to index the string
loop:
	lda	.string,x	; Load character from string into A reg
	beq	end		; If the character was 0, jump to end label
	jsr	CHROUT		; Output character stored in A register
	inx			; Increment X register
	jmp	loop		; Jump back to loop label to print next char
end:
	jsr	CHRIN		; Read input until Enter/Return is pressed
	rts			; Return to caller

.string !pet	"hello, world!!",13,0


src/5greenhearts.asm
; *=$0801
; !byte $01,$08,$0b,$08,$01,$00,$9e,$32,$30,$36,$31,$00,$00,$00
!cpu 65c02		; set target processor
*=$0801				; Assembled code should start at $0801
				; (where BASIC programs start)
				; The real program starts at $0810 = 2064
!byte $0C,$08			; $080C - pointer to next line of BASIC code
!byte $0A,$00			; 2-byte line number ($000A = 10)
!byte $9E			; SYS BASIC token
!byte $20			; [space]
!byte $32,$30,$36,$34		; $32="2",$30="0",$36="6",$34="4"
				; (ASCII encoded nums for dec starting addr)
!byte $00			; End of Line
!byte $00,$00			; This is address $080C containing
				; 2-byte pointer to next line of BASIC code
				; ($0000 = end of program)
*=$0810				; Here starts the real program

LDA #0
STA $9F25	; Select primary VRAM address
LDA #$20	; VPOKE 1st argument (The 0x00 in this is the 0 bank)
STA $9F22	; Set primary address bank to 0, stride to 2

; VPOKE 0,0,2
; VPOKE 0,1,8
; The following is the same as the 2 above VPOKE statements

LDA #0		; VPOKE 2nd argument
STA $9F20	; Set Primary address low byte to 0
LDA #0		; Not using the high byte, just want to stay on <0,0>
STA $9F21	; Set primary address high byte to 0
LDA #2		; VPOKE 3rd argument (set the character to "B")
STA $9F23	; Writing $73 to primary address ($00:$0000)

; Set the color to orange
LDA #1		; VPOKE 2nd argument (next byte over)
STA $9f20	; Next byte over
LDA #8		; VPOKE 3rd argument (orange color code)
STA $9f23	; Write the color
rts

src/videoramtest.asm
*=$0801
!byte $01,$08,$0b,$08,$01,$00,$9e,$32,$30,$36,$31,$00,$00,$00

!cpu 65c02		; set target processor
*=$0801				; Assembled code should start at $0801
				; (where BASIC programs start)
				; The real program starts at $0810 = 2064
!byte $0C,$08			; $080C - pointer to next line of BASIC code
!byte $0A,$00			; 2-byte line number ($000A = 10)
!byte $9E			; SYS BASIC token
!byte $20			; [space]
!byte $32,$30,$36,$34		; $32="2",$30="0",$36="6",$34="4"
				; (ASCII encoded nums for dec starting addr)
!byte $00			; End of Line
!byte $00,$00			; This is address $080C containing
				; 2-byte pointer to next line of BASIC code
				; ($0000 = end of program)
*=$0810				; Here starts the real program

; LDA #0
; STA $9F25	; Select primary VRAM address
; LDA #$20	; VPOKE 1st argument (The 0x00 in this is the 0 bank)
; STA $9F22	; Set primary address bank to 0, stride to 2

; VPOKE 0,0,2
; VPOKE 0,1,8
; The following is the same as the 2 above VPOKE statements

LDA #0		; VPOKE 2nd argument
STA $9F20	; Set Primary address low byte to 0
LDA #0		; Not using the high byte, just want to stay on <0,0>
STA $9F21	; Set primary address high byte to 0
LDA #2		; VPOKE 3rd argument (set the character to "B")
STA $9F23	; Writing $73 to primary address ($00:$0000)

; Set the color to orange
LDA #1		; VPOKE 2nd argument (next byte over)
STA $9f20	; Next byte over
LDA #8		; VPOKE 3rd argument (orange color code)
STA $9f23	; Write the color
rts

src/helloworldretro.65c02.asm
; Hello world example
; https://github.com/JimmyDansbo/cx16stuff/blob/master/helloworld.asm
; this will compile with retro assembler (RA)
; a BASIC header
*=$0801			; Assembled code should start at $0801
				; (where BASIC programs start)
				; The real program starts at $0810 = 2064
				; .byte RA directive
.byte $0C,$08		; $080C - pointer to next line of BASIC code
.byte $0A,$00		; 2-byte line number ($000A = 10)
.byte $9E			; SYS BASIC token
.byte $20			; [space]
.byte $32,$30,$36,$34	; $32="2",$30="0",$36="6",$34="4" = 2064
					; (ASCII encoded nums for dec starting addr)
.byte $00			; End of Line
.byte $00,$00		; This is address $080C containing
					; 2-byte pointer to next line of BASIC code
					; ($0000 = end of program)
*=$0810				; Here starts the real program

CHROUT=$FFD2		; CHROUT outputs a character (C64 Kernal API)
CHRIN=$FFCF			; CHRIN read from default input

	ldx	#0			; X register is used to index the string
output
	lda	message,x	; Load character from string into A reg
	beq	exit		; If the character was 0, jump to exit label
	jsr	CHROUT		; Output character stored in A register
	inx				; Increment X register
	jmp	output		; Jump back to loop label to print next char

exit
	jsr	CHRIN		; Read input until Enter/Return is pressed
	rts			; Return to caller

message
	.textz "HELLO, WORLD!!" ; .textz RA directive a text ending with a zero byte


src/helloworldretroWithCLS.65c02.asm
; Hello world example
; https://github.com/JimmyDansbo/cx16stuff/blob/master/helloworld.asm
; this will compile with retro assembler (RA)
; a BASIC header
*=$0801			; Assembled code should start at $0801
				; (where BASIC programs start)
				; The real program starts at $0810 = 2064
				; .byte RA directive
.byte $0C,$08		; $080C - pointer to next line of BASIC code
.byte $0A,$00		; 2-byte line number ($000A = 10)
.byte $9E			; SYS BASIC token
.byte $20			; [space]
.byte $32,$30,$36,$34	; $32="2",$30="0",$36="6",$34="4" = 2064
					; (ASCII encoded nums for dec starting addr)
.byte $00			; End of Line
.byte $00,$00		; This is address $080C containing
					; 2-byte pointer to next line of BASIC code
					; ($0000 = end of program)
*=$0810				; Here starts the real program

CHROUT=$FFD2		; CHROUT outputs a character (C64 Kernal API)
CHRIN=$FFCF			; CHRIN read from default input
					; ref https://github.com/X16Community/x16-docs/blob/master/X16%20Reference%20-%2004%20-%20KERNAL.md#function-name-graph_clear

clrscreen:
					; C64 colors https://www.c64-wiki.com/wiki/Color
	lda	#$0D		; 10 White background / Black text
					; 01  Black background /  White text
					; 05  Black background /  green text
					; 0D  Black background /  light green text
	sta	$0376
	lda	#147		; Clear screen
	jsr	CHROUT		; CHROUT outputs a character (C64 Kernal API)\
	
main:
	ldx	#0			; X register is used to index the string
output
	lda	message,x	; Load character from string into A reg
	beq	exit		; If the character was 0, jump to exit label
	jsr	CHROUT		; Output character stored in A register
	inx				; Increment X register
	jmp	output		; Jump back to loop label to print next char

exit:
	jsr	CHRIN		; Read input until Enter/Return is pressed
	rts			; Return to caller

message
	.textz "HELLO, WORLD!!" ; .textz RA directive a text ending with a zero byte
